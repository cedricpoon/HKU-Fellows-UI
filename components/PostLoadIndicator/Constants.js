import { placeholder } from 'hkufui/theme/palette';

export const LOADER_GROUP_HEIGHT = 80;

export const FADE_IN_DURATION = 40;

export const EST_HEADER_HEIGHT = 65;

export const EST_FOOTER_HEIGHT = 55;

export const PLACEHOLDER_DURATION = 1000;

export const LIGHTER_COLOR = placeholder.lighter;

export const DARKER_COLOR = placeholder.darker;
