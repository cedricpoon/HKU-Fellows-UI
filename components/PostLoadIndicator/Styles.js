import { StyleSheet } from 'react-native';
import { gridBase, comment } from 'hkufui/theme/grid';
import { seperator } from 'hkufui/theme/palette';

const hrBase = {
  height: 1,
  backgroundColor: seperator,
};

const placeholderBase = {
  height: gridBase * 2,
  alignSelf: 'flex-start',
  justifyContent: 'center',
  backgroundColor: seperator,
  borderRadius: gridBase / 2
}

export default StyleSheet.create({
  placeholderGroup: {
    marginLeft: gridBase * 7,
    marginTop: gridBase,
    marginBottom: gridBase * 2 + 1.5 /* 1.5 as complement to actual 82.5 height */,
    marginRight: gridBase * 4
  },
  paragraphGroup: {
    marginHorizontal: comment.spacingHorizontal,
    marginTop: comment.spacingVertical + gridBase
  },
  placeholder: {
    ...placeholderBase,
    marginTop: gridBase * 2
  },
  placeholderNoMarginTop: {
    ...placeholderBase
  },
  placeholderShort: {
    width: '30%'
  },
  placeholderLong: {
    width: '100%'
  },
  placeholderMedium: {
    width: '50%'
  },
  gradient: {
    flex: 1,
    width: gridBase * 25
  },
  hr: {
    ...hrBase,
    marginTop: gridBase,
    marginLeft: gridBase * 7,
    width: '100%',
  },
  headlineHr: {
    ...hrBase,
    marginHorizontal: comment.spacingHorizontal,
    marginTop: comment.spacingVertical + gridBase
  }
});
