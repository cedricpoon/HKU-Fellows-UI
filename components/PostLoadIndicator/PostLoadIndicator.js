import React, { PureComponent } from 'react';
import { Dimensions } from 'react-native';
import PropTypes from 'prop-types';
import { Content, View } from 'native-base';
import { PlaceholderContainer, Placeholder } from 'react-native-loading-placeholder';
import LinearGradient from 'react-native-linear-gradient';

import BarIndicator from './BarIndicator/BarIndicator';

import styles from './Styles';
import * as consts from './Constants';

class PostLoadIndicator extends PureComponent {

  _renderGradient() {
    return (
      <LinearGradient
        colors={[consts.LIGHTER_COLOR, consts.DARKER_COLOR, consts.LIGHTER_COLOR]}
        start={{ x: 1.0, y: 0.0 }}
        end={{ x: 0.0, y: 0.0 }}
        style={styles.gradient}
      />
    );
  }

  render() {
    const { multiple, page, ...restProps } = this.props;
    const { height } = Dimensions.get('window');
    let context = [];
    const groupHeight = page ? consts.PARAGRAPH_GROUP_HEIGHT : consts.LOADER_GROUP_HEIGHT;

    if (page || multiple) {
      if (page) {
        // render headline
        context.push(
          <View key={-1}>
            <Placeholder
              style={[styles.placeholder, styles.placeholderMedium, styles.paragraphGroup]}
            />
            <View
              style={styles.headlineHr}
            />
          </View>
        );
      }
      for (let i = 0; i < height - consts.EST_FOOTER_HEIGHT - consts.EST_HEADER_HEIGHT; i += groupHeight) {
        context.push(
          <BarIndicator
            key={i}
            delay={(i / consts.LOADER_GROUP_HEIGHT + 1) * consts.DELAY_BASE}
            paragraph={page}
            {...restProps}
          />
        );
      }
    } else {
      context.push(
        <BarIndicator key={0} {...restProps} />
      );
    }

    return(
      <Content scrollEnabled={false}>
        <PlaceholderContainer
          animatedComponent={this._renderGradient()}
          duration={consts.PLACEHOLDER_DURATION}
          delay={consts.PLACEHOLDER_DURATION}
        >
          {context}
        </PlaceholderContainer>
      </Content>
    );
  }
}

PostLoadIndicator.propTypes = {
  multiple: PropTypes.bool,
  page: PropTypes.bool
}

export default PostLoadIndicator;
