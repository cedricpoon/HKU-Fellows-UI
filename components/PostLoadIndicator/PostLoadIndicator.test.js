import React from 'react';
import { shallow } from 'enzyme';

import PostLoadIndicator from './PostLoadIndicator';

describe('Testing PostLoadIndicator Component', () => {
  it('renders as expected', () => {
    const wrapper = shallow(
      <PostLoadIndicator />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders as expected in multiple', () => {
    const wrapper = shallow(
      <PostLoadIndicator multiple />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders as a page', () => {
    const wrapper = shallow(
      <PostLoadIndicator page />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
