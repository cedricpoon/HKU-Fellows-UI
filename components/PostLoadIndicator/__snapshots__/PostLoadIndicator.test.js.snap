// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing PostLoadIndicator Component renders as a page 1`] = `
<Styled(Content)
  scrollEnabled={false}
>
  <PlaceholderContainer
    animatedComponent={
      <LinearGradient
        colors={
          Array [
            "#eee",
            "#ddd",
            "#eee",
          ]
        }
        end={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        start={
          Object {
            "x": 1,
            "y": 0,
          }
        }
        style={
          Object {
            "flex": 1,
            "width": 125,
          }
        }
      />
    }
    delay={1000}
    duration={1000}
  >
    <Styled(ViewNB)
      key="-1"
    >
      <Placeholder
        style={
          Array [
            Object {
              "alignSelf": "flex-start",
              "backgroundColor": "#eee",
              "borderRadius": 2.5,
              "height": 10,
              "justifyContent": "center",
              "marginTop": 10,
            },
            Object {
              "width": "50%",
            },
            Object {
              "marginHorizontal": 15,
              "marginTop": 15,
            },
          ]
        }
      />
      <Styled(ViewNB)
        style={
          Object {
            "backgroundColor": "#eee",
            "height": 1,
            "marginHorizontal": 15,
            "marginTop": 15,
          }
        }
      />
    </Styled(ViewNB)>
    <BarIndicator
      delay={40}
      key="0"
      paragraph={true}
    />
    <BarIndicator
      delay={92.5}
      key="105"
      paragraph={true}
    />
    <BarIndicator
      delay={145}
      key="210"
      paragraph={true}
    />
    <BarIndicator
      delay={197.5}
      key="315"
      paragraph={true}
    />
    <BarIndicator
      delay={250}
      key="420"
      paragraph={true}
    />
    <BarIndicator
      delay={302.5}
      key="525"
      paragraph={true}
    />
    <BarIndicator
      delay={355}
      key="630"
      paragraph={true}
    />
    <BarIndicator
      delay={407.5}
      key="735"
      paragraph={true}
    />
    <BarIndicator
      delay={460}
      key="840"
      paragraph={true}
    />
    <BarIndicator
      delay={512.5}
      key="945"
      paragraph={true}
    />
    <BarIndicator
      delay={565}
      key="1050"
      paragraph={true}
    />
    <BarIndicator
      delay={617.5}
      key="1155"
      paragraph={true}
    />
  </PlaceholderContainer>
</Styled(Content)>
`;

exports[`Testing PostLoadIndicator Component renders as expected 1`] = `
<Styled(Content)
  scrollEnabled={false}
>
  <PlaceholderContainer
    animatedComponent={
      <LinearGradient
        colors={
          Array [
            "#eee",
            "#ddd",
            "#eee",
          ]
        }
        end={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        start={
          Object {
            "x": 1,
            "y": 0,
          }
        }
        style={
          Object {
            "flex": 1,
            "width": 125,
          }
        }
      />
    }
    delay={1000}
    duration={1000}
  >
    <BarIndicator
      delay={0}
      key="0"
    />
  </PlaceholderContainer>
</Styled(Content)>
`;

exports[`Testing PostLoadIndicator Component renders as expected in multiple 1`] = `
<Styled(Content)
  scrollEnabled={false}
>
  <PlaceholderContainer
    animatedComponent={
      <LinearGradient
        colors={
          Array [
            "#eee",
            "#ddd",
            "#eee",
          ]
        }
        end={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        start={
          Object {
            "x": 1,
            "y": 0,
          }
        }
        style={
          Object {
            "flex": 1,
            "width": 125,
          }
        }
      />
    }
    delay={1000}
    duration={1000}
  >
    <BarIndicator
      delay={40}
      key="0"
    />
    <BarIndicator
      delay={80}
      key="80"
    />
    <BarIndicator
      delay={120}
      key="160"
    />
    <BarIndicator
      delay={160}
      key="240"
    />
    <BarIndicator
      delay={200}
      key="320"
    />
    <BarIndicator
      delay={240}
      key="400"
    />
    <BarIndicator
      delay={280}
      key="480"
    />
    <BarIndicator
      delay={320}
      key="560"
    />
    <BarIndicator
      delay={360}
      key="640"
    />
    <BarIndicator
      delay={400}
      key="720"
    />
    <BarIndicator
      delay={440}
      key="800"
    />
    <BarIndicator
      delay={480}
      key="880"
    />
    <BarIndicator
      delay={520}
      key="960"
    />
    <BarIndicator
      delay={560}
      key="1040"
    />
    <BarIndicator
      delay={600}
      key="1120"
    />
    <BarIndicator
      delay={640}
      key="1200"
    />
  </PlaceholderContainer>
</Styled(Content)>
`;
